src/uct/tcp/tcp_cm.c

```diff
@@ -8,100 +8,299 @@
+    ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr,
+                     UCS_SOCKADDR_STRING_LEN);
+
+    ucs_debug("tcp_ep %p: %s -> %s for the [%s]<->[%s] connection",
+              ep, uct_tcp_ep_cm_state[old_conn_state].name,
+              uct_tcp_ep_cm_state[ep->conn_state].name,
+              str_local_addr, str_remote_addr);
+}
+
+static ucs_status_t uct_tcp_cm_send_conn_req(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface         = ucs_derived_of(ep->super.super.iface,
+                                                    uct_tcp_iface_t);
+    uct_tcp_cm_conn_pkt_t conn_pkt = {
+        .event                     = UCT_TCP_CM_CONN_REQ
+    };
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    size_t addr_size;
+    ucs_status_t status;
+
+    status = ucs_sockaddr_sizeof((const struct sockaddr*)&iface->config.ifaddr,
+                                 &addr_size);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    memcpy(&conn_pkt.iface_addr, &iface->config.ifaddr, addr_size);
+
+    status = ucs_socket_send(ep->fd, &conn_pkt, sizeof(conn_pkt));
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ucs_debug("tcp_ep %p: conection request sent to %s",
+              ep, ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr,
+                                   UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_cm_recv_conn_req(uct_tcp_ep_t *ep,
+                                             struct sockaddr *peer_addr)
+{
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    uct_tcp_cm_conn_pkt_t conn_pkt;
+    ucs_status_t status;
+
+    status = ucs_socket_recv(ep->fd, &conn_pkt, sizeof(conn_pkt));
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ucs_assertv(conn_pkt.event == UCT_TCP_CM_CONN_REQ, "ep=%p", ep);
+    *peer_addr = conn_pkt.iface_addr;
+
+    ucs_debug("tcp_ep %p: received the connection request from %s",
+              ep, ucs_sockaddr_str(peer_addr, str_remote_addr,
+                                   UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_cm_send_conn_ack(uct_tcp_ep_t *ep)
+{
+    uct_tcp_cm_conn_pkt_t conn_pkt = {
+        .event                     = UCT_TCP_CM_CONN_ACK
+    };
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+
+    status = ucs_socket_send(ep->fd, &conn_pkt, sizeof(conn_pkt));
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ucs_debug("tcp_ep %p: conection ack sent to %s",
+              ep, ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr,
+                                   UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_cn_recv_conn_ack(uct_tcp_ep_t *ep)
+{
+    uct_tcp_cm_conn_pkt_t conn_pkt;
+    ucs_status_t status;
+
+    status = ucs_socket_recv(ep->fd, &conn_pkt, sizeof(conn_pkt));
+
+    ucs_assertv((status != UCS_OK) ||
```

Comment:
what does this check? Why do we expect non-ok status?

Reply:
We can do:
```
if (status == UCS_OK) {
    ucs_assertv((conn_pkt.event == UCT_TCP_CM_CONN_ACK), "ep=%p", ep);
}
```

but it requires an additional check for the status.



> Why do we expect non-ok status?

It expects that either (status != UCS_OK) or (event is CONN_ACK).